Endpoints for vocabulary (#176)
resource-watch/doc-api: source/includes/api/_vocabulary.md

> # Vocabularies and tags

Reviewer: Looks like there are line break issues throughout here. They don't impact the rendered markdown. If you can think of an easy way to fix them (regex replace (\w)\n(\w) with $1 $2 perhaps?) then please go ahead but otherwise don't worry about it.

Author: Seems like a recent change in my IDE made it hard-wrap lines at 120 characters. If it's not causing any visible issues, I'd rather leave it as it is.

exist to improve and empower your use cases of those resources, so it should only be used by applications and users that
are already comfortable with the basics, and want to take their knowledge of the RW API to the next level.

The behavior of vocabulary and tags endpoints aims to be, as much as possible, independent from the type of the target
resource it references. In the detailed endpoint documentation below we'll cover the different endpoints in depth, and
highlight the differences in behavior when handling different resource types, but you can safely assume that, for most
of it, the behavior described for a type of resource will be the same for all 3 types.

### A note on resources

> Creating a new vocabulary/tag for a widget

```shell
curl -X POST https://api.resourcewatch.org/v1/dataset/AAA/widget/BBB/vocabulary/VVV \
-H "Content-Type: application/json"  -d \
 '{
   "application": <application>,
   "tags": [<tags>]
  }'
```

In the context of the vocabulary/tag service, widgets and layers are identified not only by their own id but also by the
id you specify as being their associated dataset. Using the side example as reference, this API request would create a

> vocabulary `VVV` for the resource widget `BBB` associated with dataset `AAA`. If you later reference the same widget id,

Reviewer: Does this ever actually happen? Like, is this a well-defined use case?

Author: No, this should never happen, AFAIK. Layers and widgets should only belong to a single dataset, so these types of cases should probably issue an error rather than have this unexpected behavior. But as that's not the case, I preferred to give readers a heads up. I will clarify that this is unintentional behavior

Reviewer: 👍

different vocabularies and tags.

## The special `knowledge_graph` vocabulary

Throughout the whole vocabulary/tags service, there is a special vocabulary named `knowledge_graph`. When used, this
vocabulary acts as any other vocabulary but has additional logic built-in for the purpose of integrating vocabulary/tags
with the [graph](#graph) available as part of the RW API.

Whenever you create or clone a vocabulary named `knowledge_graph` for a resource,
the [dataset](#creating-dataset-graph-nodes), [layer](#creating-layer-graph-nodes)
or [widget](#creating-widget-graph-nodes) node is added to the graph if not already present, and
the [tags are associated to it](#associating-concepts-to-graph-nodes). Updating the `knowledge_graph` vocabulary also
updates the corresponding [tags associated to it](#updating-concepts-associated-with-graph-nodes) on the graph side, and
deleting the vocabulary
also [deletes the corresponding tags in the graph](#deleting-concepts-associated-with-graph-nodes).

For more details on the graph, its purpose, and how to use it, please refer to the dedicated documentation section on
the [graph](#graph) and [its endpoints](#graph7).

> ## Associations between vocabularies, tags and resources

Reviewer: I think I'm a little confused about the use case now...if vocabularies are defined by their association with individual resources, then what's the point of having a vocabulary at all (as opposed to just a list of tags)?

The concept docs made it seem like vocabularies were conceptual groupings of tags that could be reused across resources, but now it seems like maybe that's not the case?

Author: this is a lot harder and more complex to explain over text, please ping me on slack so we can jump on a call

create. Each of these keys must have an object-type value, which in turn must contain the following keys:

- `tags`: an array of tags
- `application`: the application associated with the vocabulary.

You can create different vocabularies for different applications in a single request.

Assuming the creation was successful, the response will contain a list of all vocabularies and their respective tags
associated with the specified resource, for all applications.

If you want to create new vocabulary/tags for your resources, you must have the necessary user account permissions.
Specifically:

- the user must be logged in and belong to the same application as the resource
- the user must match one of the following:
    - have role `ADMIN`
    - have role `MANAGER` and be the resource's owner (through the `userId` field of the resource)

When creating vocabularies/tags for a resource, the dataset id specified in the URL is validated, and the requests fail
if a dataset with the given id does not exist. When creating a vocabulary for a widget or layer, you should ensure you

> specify the dataset id that matches that of the widget/layer, as that validation is not done automatically - this is a

Reviewer: I would make this bold or put it in an info box or something - seems important to make sure people aren't relying on this behavior

Reviewer: Same with other places that this text appears

Author: :+1:

400            | This relationship already exists | The resource already has one or more vocabularies with the same name(s) for the specified application(s)
400            | - `<value>`: `<value>` check failed. -  | The body object has a `<value>` key which content is invalid (is not an object or is missing `tags` and/or `application`).
401            | Unauthorized   | You are not authenticated.
403            | Forbidden      | You are trying to create a vocabulary for a resource whose `application` value is not associated with your user account.
404            | 404 - {\"errors\":[{\"status\":404,\"detail\":\"Dataset with id `<dataset id>` doesn't exist\"}]}      | You are trying to create a vocabulary for a dataset that doesn't exist.

## Updating vocabularies/tags

If you have already associated vocabularies and tags to your resources and would like to modify those tags, the next set
of endpoints is for you.

There are three types of endpoints for updating existing vocabulary/tags:

- endpoints that allow you to modify a single vocabulary (and its tags) for a resource
- endpoints that allow you to modify multiple vocabularies/tags for a single resource with a single request

> - endpoint that allow you to add new tags to an existing dataset vocabulary

Reviewer: ```suggestion
- endpoint that allows you to add new tags to an existing vocabulary. This endpoint only exists for dataset resources, not widgets or layers.
```
Just thought it would be good to be more explicit here, since the original wording caught me off guard.

-H "Content-Type: application/json"  -d \
 '{
   "application": <application>,
   "tags": [<tags>]
  }'
```

> Updating a vocabulary for a layer

```shell
curl -X PATCH https://api.resourcewatch.org/v1/dataset/<dataset-id>/layer/<layer-id>/vocabulary/<vocabulary-id> \
-H "Content-Type: application/json"  -d \
 '{
   "application": <application>,
   "tags": [<tags>]
  }'
```

> Example response

```json
{
    "data": [
        {
            "id": "vocabularyNameOne",
            "type": "vocabulary",
            "attributes": {
                "tags": [
                    "tag1",
                    "tag2",
                    "tag3"
                ]
            }
        },
        {
            "id": "science",
            "type": "vocabulary",
            "attributes": {
                "tags": [
                    "maths",
                    "astronomy"
                ]
            }
        },
        {
            "id": "aaa",
            "type": "vocabulary",
            "attributes": {
                "tags": [
                    "biology",
                    "chemistry"
                ]
            }
        }
    ]
}
```

> This set of endpoints will allow you to update the tags of an existing vocabulary/tags for a resource.

Reviewer: ```suggestion
This set of endpoints will allow you to overwrite the tags of an existing vocabulary/tags for a resource.
```

> Deleting a single vocabulary/tag for a dataset

```shell
curl -X DELETE https://api.resourcewatch.org/v1/dataset/<dataset-id>/vocabulary/<vocabulary-id>
```

> Deleting a single vocabulary/tag for a widget

```shell
curl -X DELETE https://api.resourcewatch.org/v1/dataset/<dataset-id>/widget/<widget-id>/vocabulary/<vocabulary-id>
```

> Deleting a single vocabulary/tag for a layer

```shell
curl -X DELETE https://api.resourcewatch.org/v1/dataset/<dataset-id>/layer/<layer-id>/vocabulary/<vocabulary-id>
```

> Example response

```json
{
    "data": [
        {
            "id": "country",
            "type": "vocabulary",
            "attributes": {
                "tags": [
                    "Spain",
                    "Italy",
                    "Portugal"
                ]
            }
        },
        {
            "id": "color",
            "type": "vocabulary",
            "attributes": {
                "tags": [
                    "red",
                    "green",
                    "blue"
                ]
            }
        }
    ]
}
```

This set of endpoints will allow you to delete a single vocabulary and its tags, for a resource.

> As a query param, you can optionally provide an `app` or `application` values, to identify the application of the

Reviewer: What will happen if you don't do this and there are two vocabularies with the same name but different applications (can that even happen?)? Will it potentially delete multiple? Will it pick one randomly?

If you want to delete vocabularies/tags for your resources, you must have the necessary user account permissions.
Specifically:

- the user must be logged in and belong to the same application as the resource
- the user must match one of the following:
    - have role `ADMIN`
    - have role `MANAGER` and be the resource's owner (through the `userId` field of the resource)

When deleting vocabulary/tags for a resource, the dataset id specified in the URL is validated, and the requests fail if
a dataset with the given id does not exist. When deleting a vocabulary for a widget or layer, you should ensure you
specify the dataset id that matches that of the widget/layer, as that validation is not done automatically - this is a
known limitation of the current implementation and may be modified at any time.

#### Errors for deleting a single vocabulary/tags for a resource

Error code     | Error message  | Description
-------------- | -------------- | --------------
401            | Unauthorized   | You are not authenticated.
403            | Forbidden      | You are trying to delete a vocabulary for a resource whose `application` value is not associated with your user account.

> 404            | 404 - {\"errors\":[{\"status\":404,\"detail\":\"Dataset with id `<dataset id>` doesn't exist\"}]}      | You are trying to create a vocabulary for a dataset that doesn't exist.

Reviewer: ```suggestion
404            | 404 - {\"errors\":[{\"status\":404,\"detail\":\"Dataset with id `<dataset id>` doesn't exist\"}]}      | You are trying to delete a vocabulary for a dataset that doesn't exist.
```

404            | Relationship between `<vocabulary id> and dataset - `<dataset id> and dataset: `<dataset id> doesn't exist      | You are trying to delete a vocabulary that doesn't exist.

### Deleting all vocabulary/tags for a resource

> Deleting all vocabularies/tag for a dataset

```shell

> curl -X DELETE https://api.resourcewatch.org/v1/dataset/<dataset-id>/vocabulary/<vocabulary-id>

Reviewer: Confused...this is identical to the single deletion case

Reviewer: Along with the other examples in this section

Author: copy-pasting gone wrong. will fix it

```

> Deleting all vocabularies/tag for a widget

```shell
curl -X DELETE https://api.resourcewatch.org/v1/dataset/<dataset-id>/widget/<widget-id>/vocabulary/<vocabulary-id>
```

> Deleting all vocabularies/tag for a layer

```shell
curl -X DELETE https://api.resourcewatch.org/v1/dataset/<dataset-id>/layer/<layer-id>/vocabulary/<vocabulary-id>
```

> Example response

```json
{
    "data": [
        {
            "id": "country",
            "type": "vocabulary",
            "attributes": {
                "tags": [
                    "Spain",
                    "Italy",
                    "Portugal"
                ]
            }
        },
        {
            "id": "color",
            "type": "vocabulary",
            "attributes": {
                "tags": [
                    "red",
                    "green",
                    "blue"
                ]
            }
        }
    ]
}
```

**Know issue warning** There's a known issue where you may experience

> a `404 - Relationship between undefined and dataset - <dataset id> and dataset: <dataset id> doesn't exist` error when

Reviewer: 😬

If you want to delete vocabularies/tags for your resources, you must have the necessary user account permissions.
Specifically:

- the user must be logged in and belong to the same application as the resource
- the user must match one of the following:
    - have role `ADMIN`
    - have role `MANAGER` and be the resource's owner (through the `userId` field of the resource)

When deleting vocabulary/tags for a resource, the dataset id specified in the URL is validated, and the requests fail if
a dataset with the given id does not exist. When deleting vocabularies for a widget or layer, you should ensure you
specify the dataset id that matches that of the widget/layer, as that validation is not done automatically - this is a
known limitation of the current implementation and may be modified at any time.

#### Errors for deleting all vocabularies/tags for a resource

Error code     | Error message  | Description
-------------- | -------------- | --------------
401            | Unauthorized   | You are not authenticated.
403            | Forbidden      | You are trying to delete vocabularies for a resource whose `application` value is not associated with your user account.

> 404            | 404 - {\"errors\":[{\"status\":404,\"detail\":\"Dataset with id `<dataset id>` doesn't exist\"}]}      | You are trying to create a vocabulary for a dataset that doesn't exist.

Reviewer: ```suggestion
404            | 404 - {\"errors\":[{\"status\":404,\"detail\":\"Dataset with id `<dataset id>` doesn't exist\"}]}      | You are trying to delete vocabularies for a dataset that doesn't exist.
```

                        "dataset": "7fa6ec77-5ab0-43f4-9a4c-a3d19bed1e90",
                        "type": "layer"
                    },
                    {
                        "tags": [
                            "tag1",
                            "tag2"
                        ],
                        "id": "0002ed77-f07d-4231-b4eb-cfda77eeafe5",
                        "dataset": "7fa6ec77-5ab0-43f4-9a4c-a3d19bed1e91",
                        "type": "layer"
                    }
                ],
                "name": "test1",
                "application": "rw"
            }
        },
        {
            "id": "forestChange",
            "type": "vocabulary",
            "attributes": {
                "resources": [],
                "name": "forestChange",
                "application": "gfw"
            }
        }
    ]
}
```

This endpoint will give you a list of vocabularies. Inside each, you'll find a list of resources, identifying the
resource id, type, associated dataset (if the resource is a dataset, the dataset id will be equal to the resource id)and
the tags that said vocabulary associates to that resource

> You can optionally pass a `limit` integer value as a query parameter if you wish to limit the number of vocabularies

Reviewer: Is this paginated? If not let's call that out explicitly

Author: no, it just supports a simple "limit" parameter.

returned by calls to this endpoint

### Getting a single vocabulary

> Getting a single vocabulary, all its resources and tags

```shell

> curl -X GET https://api.resourcewatch.org/v1/vocabulary/<vocabulary id>

Reviewer: Replace all `vocabulary id` with `vocabulary-id` for consistency

equal to the resource id) and the tags that said vocabulary associates to that resource.

Keep in mind that there's an implicit application filter applied to the data returned - only vocabularies and
associations for the corresponding application are returned. By default, this filter uses the `rw` value, but you can
modify it with the `application` or `app` query parameters.

### Getting tags for single vocabulary

> Getting the tags for a single vocabulary

```shell
curl -X GET https://api.resourcewatch.org/v1/vocabulary/<vocabulary id>/tags
```

> Example response

```json
{
    "data": [
        "tag11",
        "tag21",
        "tag1",
        "tag2"
    ]
}
```

This endpoint will give you a flat list of all the tags associated with the specified vocabulary.

Keep in mind that there's an implicit application filter applied to the data returned - only tags for the corresponding
application are returned. By default, this filter uses the `rw` value, but you can modify it with the `application`
or `app` query parameters.

## Getting resources by vocabulary and tag

> Getting datasets by vocabulary and tag

```shell

> curl -X GET https://api.resourcewatch.org/v1/dataset/vocabulary/find

Reviewer: Should add an example query param here

### Vocabularies and resources

Despite not being 100% accurate, it's useful to think that a vocabulary is uniquely identified by a tuple of 3 values:

- The resource (id and type) to which it's associated
- Its `name` or `id` (these are 2 names for the same underlying value)
- Its `application`

Putting it in other words: you can think of a vocabulary as being **associated with only one resource at a time**. So,
if you create, update or delete a vocabulary, you are only doing so *for that resource*. You can safely modify the
vocabularies for a resource without fear of affecting vocabularies and tags for other resources.

Further down, we'll cover endpoints that allow you to discover resources by performing searches based on vocabularies
and tags. These endpoints are able to aggregate these different vocabularies and tags from multiple resources, and
produce a unified response - making it look like vocabularies are shared across resources, and thus producing meaningful
results.

> If you find the above slightly confusing, we advice you explore these docs and learn more about the different endpoints available - maybe even try to add vocabularies and tags to one of your resources - and then come back and see if your newly gained knowledge helps you get a deeper understanding of how vocabularies and tags work on the RW API. 

Reviewer: ```suggestion
If you find the above slightly confusing, we advise you explore these docs and learn more about the different endpoints available - maybe even try to add vocabularies and tags to one of your resources - and then come back and see if your newly gained knowledge helps you get a deeper understanding of how vocabularies and tags work on the RW API. 
```
